docs(grid): enhance expando kb (#1761)
telerik/blazor-docs: knowledge-base/grid-binding-to-expando-object.md

## Notes

> The following list contains information about requirements and limitations when binding the Grid to `ExpandoObject`. They are related to using nullable data, editing, filtering, grouping, and nested model properties. A [runnable example](#example) is available afterwards.

Reviewer: Afterwards typically refers to time rather than a specific position within a document.
```suggestion
The following list contains information about requirements and limitations when binding the Grid to `ExpandoObject`. They are related to using nullable data, editing, filtering, grouping, and nested model properties. A [runnable example](#example) is available later in the article.
```

### Editing

> * If any non-string property in the `ExpandoObject` is `nullable`, then do not set [`EditorType`]({%slug components/grid/editing/overview%}#customize-the-editor-fields) for the column. Instead, use a [Grid column `EditorTemplate`]({%slug grid-templates-editor%}). The following code snippet is part of the full example below.

Reviewer: Don't introduce a bullet list with a single item.

<div class="skip-repl"></div>

````CSHTML
<GridColumn Field="PropertyDate"
            FieldType="@GetPropertyType("PropertyDate")"
            Title="Date declared">
    <EditorTemplate>
        @{
            var editItem = (IDictionary<string, object>)context;
            DateTime? dateValue = (DateTime?)(editItem["PropertyDate"]);
            <TelerikDatePicker Value="@dateValue"
                                ValueChanged="@( (DateTime? newValue) => editItem["PropertyDate"] = newValue )"
                                DebounceDelay="0" />
        }
    </EditorTemplate>
</GridColumn>
````

### Filtering

* Filtering in nullable data requires you to use [`OnRead` data binding]({%slug common-features-data-binding-onread%}). Populate the missing `MemberType` properties in the `CompositeFilterDescriptor`s in the [`OnRead` event argument]({%slug common-features-data-binding-onread%}#event-argument), namely in `args.Request.Filters`. Also see the [related public item](https://feedback.telerik.com/blazor/1499221-incorrect-filter-operators-when-the-fieldtype-is-set-to-a-nullable-numeric-data-type). The following code snippet is part of the full example below.

> <div class="skip-repl"></div>

Reviewer: Indent the entire code snippet so that it gets rendered as nested under the bullet item above.

        .Each(x =>
        {
            x.FilterDescriptors.OfType<FilterDescriptor>()
                .Each(y => y.MemberType = GetPropertyType(y.Member, true));
        });

    // ...
}
````

* [Row filtering]({%slug grid-filter-row%}) in nullable data requires [`FilterCellTemplate`]({%slug grid-templates-filter%}) for numeric and boolean properties. See columns `PropertyInt` and `PropertyBool` in the example below.

* [`CheckboxList` filtering]({%slug grid-checklist-filter%}) in non-nullable data requires `OnRead` data binding and the same approach as in the first bullet item above.

* The built-in `CheckboxList` filtering doesn't support nullable data. Instead, use the default [filter menu]({%slug grid-filter-menu%}) or [filter row]({%slug grid-filter-row%}). Using a filter menu template with a custom checkbox list is possible.

* `CheckboxList` filtering in grouped data with `OnRead` event requires a [`FilterMenuTemplate`]({%slug grid-templates-filter%}). You can use a [`TelerikCheckBoxListFilter` component with custom data]({%slug grid-checklist-filter%}#custom-data). The following code snippet is part of the full example below. See column `PropertyBool`.

> <div class="skip-repl"></div>

Reviewer: Again, I suggest indenting the code snippet and all other nested snippets that follow.

                            @bind-FilterDescriptor="@context.FilterDescriptor" />
````
### Grouping

* [Loading groups on demand]({%slug grid-group-lod%}) requires you to bind the Grid with [`OnRead` event]({%slug common-features-data-binding-onread%}). Then, set the `MemberType` of the group-related `FilterDescriptor` in the `OnRead` handler. Otherwise the expanded group will show only one item. Also see the [related public item for built-in support](https://feedback.telerik.com/blazor/1535071-loadgroupsondemand-does-not-let-you-group-with-an-expandoobject-as-data-source). The following code snippet is part of the full example below.

<div class="skip-repl"></div>

````CS
private async Task OnGridRead(GridReadEventArgs args)
{
    args.Request.Filters.OfType<FilterDescriptor>()
        .Each(x => x.MemberType = GetPropertyType(x.Member, true));

    // ...
}
````

* [Loading groups on demand doesn't support grouping by nullable property](https://feedback.telerik.com/blazor/1543814-loading-groups-on-demand-in-a-column-with-a-nullable-data-type-groups-all-records-under-the-null-group).

### Nested Properties

> * You can use [nested Grid model properties]({%slug grid-use-navigation-properties%}) that are `ExpandoObject`s. However, data operations like filtering, sorting and grouping are not supported for those nested properties. [Disable these features per column]({%slug components/grid/columns/bound%}#data-operations) if they are enabled for the Grid.

Reviewer: ```suggestion
* You can use [nested Grid model properties]({%slug grid-use-navigation-properties%}) that are `ExpandoObject`s. However, data operations like filtering, sorting, and grouping are not supported for those nested properties. [Disable these features per column]({%slug components/grid/columns/bound%}#data-operations) if they are enabled for the Grid.
```

### Example

> The following example demonstrates how the Grid can edit, filter and group nullable `ExpandoObject` data. The example covers multiple different scenarios at the same time, so remove any code and settings that you don't need.

Reviewer: ```suggestion
The following example demonstrates how the Grid can edit, filter, and group nullable `ExpandoObject` data. The example covers multiple different scenarios at the same time, so remove any code and settings that you don't need.
```

> >caption Using ExpandoObject with nullable data, CheckBoxList and LoadGroupsOnDemand

Reviewer: ```suggestion
>caption Using ExpandoObject with nullable data, CheckBoxList, and LoadGroupsOnDemand
```
